ledger/eval_test.go

```diff
@@ -220,3 +220,97 @@ func TestRekeying(t *testing.T) {
 
 	// TODO: More tests
 }
+
+func TestPrepareAppEvaluators(t *testing.T) {
+	eval := BlockEvaluator{
+		prevHeader: bookkeeping.BlockHeader{
+			TimeStamp: 1234,
+			Round:     2345,
+		},
+		proto: config.ConsensusParams{
+			Application: true,
+		},
+	}
+
+	// Create some sample transactions
+	payment := transactions.SignedTxnWithAD{
+		SignedTxn: transactions.SignedTxn{
+			Txn: transactions.Transaction{
+				Type: protocol.PaymentTx,
+				Header: transactions.Header{
+					Sender: basics.Address{1, 2, 3, 4},
+				},
+				PaymentTxnFields: transactions.PaymentTxnFields{
+					Receiver: basics.Address{4, 3, 2, 1},
+					Amount:   basics.MicroAlgos{Raw: 100},
+				},
+			},
+		},
+	}
+
+	appcall1 := transactions.SignedTxnWithAD{
+		SignedTxn: transactions.SignedTxn{
+			Txn: transactions.Transaction{
+				Type: protocol.ApplicationCallTx,
+				Header: transactions.Header{
+					Sender: basics.Address{1, 2, 3, 4},
+				},
+				ApplicationCallTxnFields: transactions.ApplicationCallTxnFields{
+					ApplicationID: basics.AppIndex(1),
+				},
+			},
+		},
+	}
+
+	appcall2 := appcall1
+	appcall2.SignedTxn.Txn.ApplicationCallTxnFields.ApplicationID = basics.AppIndex(2)
+
+	type evalTestCase struct {
+		group []transactions.SignedTxnWithAD
+
+		// indicates if prepareAppEvaluators should return a non-nil
+		// appTealEvaluator for the txn at index i
+		expected []bool
+	}
+
+	// Create some groups with these transactions
+	cases := []evalTestCase{
+		evalTestCase{[]transactions.SignedTxnWithAD{payment}, []bool{false}},
+		evalTestCase{[]transactions.SignedTxnWithAD{appcall1}, []bool{true}},
+		evalTestCase{[]transactions.SignedTxnWithAD{payment, payment}, []bool{false, false}},
+		evalTestCase{[]transactions.SignedTxnWithAD{appcall1, payment}, []bool{true, false}},
+		evalTestCase{[]transactions.SignedTxnWithAD{payment, appcall1}, []bool{false, true}},
+		evalTestCase{[]transactions.SignedTxnWithAD{appcall1, appcall2}, []bool{true, true}},
+		evalTestCase{[]transactions.SignedTxnWithAD{appcall1, appcall2, appcall1}, []bool{true, true, true}},
+		evalTestCase{[]transactions.SignedTxnWithAD{payment, appcall1, payment}, []bool{false, true, false}},
+		evalTestCase{[]transactions.SignedTxnWithAD{appcall1, payment, appcall2}, []bool{true, false, true}},
+	}
+
+	for _, testCase := range cases {
+		res := eval.prepareAppEvaluators(testCase.group)
```

Comment:
nit: you might use
```
t.Run(fmt.Sprintf("v=%d", v), func(t *testing.T) { ...})
```
that gives clearer idea which test fails

Reply:
done! thanks